onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+        graph.RemoveEdge(trail_cast->Index(), consumer->Index(), cast_output_index, input_index);
+        std::replace(consumer_inputs.begin(), consumer_inputs.end(), cast_output, cast_input);
+        if (producer) {
+          graph.AddEdge(producer->Index(), consumer->Index(), output_index, input_index);
+        }
+      }
+    }
+    graph.UpdateConsumerNodes(cast_input->Name(), consumers);
+  }
+  for (auto cast : casts) {
+    graph_utils::RemoveNodeOutputEdges(graph, *cast);
+    removed_nodes.push_back(cast->Index());
+  }
+  return Status::OK();
+}
+
+// RemoveBackToBackCasts
+// Remove FLOAT and FLOAT16 casts back-to-back, either FLOAT->FLOAT16 or FLOAT16->FLOAT
+// Condition: The parent cast should have only one output
+static bool RemoveBackToBackCasts(Graph& graph,
+                                  std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                  const logging::Logger& logger)
+{
+  ORT_UNUSED_PARAMETER(logger);
+  bool modified = false;
+  for (Node& node : graph.Nodes()) {
+    if (std::find(removed_nodes.begin(), removed_nodes.end(), node.Index()) == removed_nodes.end()) {
+      bool is_fp = IsCastTo(node, TensorProto::FLOAT);
+      bool is_fp16 = IsCastTo(node, TensorProto::FLOAT16);
+      if (!is_fp && ! is_fp16) {
+        continue;
+      }
+      if (node.MutableOutputDefs().size() == 1) {
+        NodeArg* cast_output = node.MutableOutputDefs()[0];
+        for (Node* child : graph.GetMutableConsumerNodes(cast_output->Name())) {
+          if (std::find(removed_nodes.begin(), removed_nodes.end(), child->Index()) == removed_nodes.end()) {
+            bool is_child_fp = IsCastTo(*child, TensorProto::FLOAT);
+            bool is_child_fp16 = IsCastTo(*child, TensorProto::FLOAT16);
+            if ((is_fp && is_child_fp16) || (is_fp16 && is_child_fp)) {
+              // The parent and child cancell out
+              VLOGS(logger, 1) << "RemoveBackToBackCasts: Removed Cast nodes  " << node.Name() << " and " << child->Name();
+              RemoveCastNodes(graph, {&node, child}, removed_nodes);
+              modified = true;
+            } else if ((is_fp16 && is_child_fp16) || (is_fp && is_child_fp)) {
+              // Child is a duplicate of parent
+              VLOGS(logger, 1) << "RemoveBackToBackCasts: Removed Cast node  " << child->Name();
+              RemoveCastNodes(graph, {child}, removed_nodes);
+              modified = true;
+            }
+          }
+        }
+      }
+    }
+  }
+  return modified;
+}
+
+// SearchUpstream:
+// Recursively DFS traverse bottom-up, the graph upstream collecting all the NodeArgs that require a cast
+// inorder to move an FP16 Cast operation up the graph.
+// Visited float NodeArgs are either in require_cast or require_type_change so that the same
+// nodearg is traversed not more than once.
+static void SearchUpstream(Graph& graph, NodeArg* node_arg,
```

Comment:
as discussed offline, we will try using Graph::ReverseDFSFrom() to traverse the graph. 
If ReverseDFSFrom is not handy enough, we will try write a DFS, just like GradientGraphBuilder::ReverseDFSWithStopEdges()

Reply:
I have removed recursion from SearchForwards and SearchBackwards.

Reply:
Please mark this as a TODO. 
I still think we should use DFS, instead of Recursive function. 

---
In reply to: [610961527](https://github.com/microsoft/onnxruntime/pull/7191#discussion_r610961527) [](ancestors = 610961527)